contracts/oracles/UniswapV3Oracle.sol

```diff
@@ -0,0 +1,143 @@
+     * @param period_ the time period
+     * @return amount_ the price of start token in quote token
+     * @return minPeriod_ the oldest period for which there is an observation in case period_ time ago there was no observation
+     */
+    function getPriceOfTokenInToken(
+        address[] memory path_,
+        uint24[] memory fees_,
+        uint128 amount_,
+        uint32 period_
+    ) public view returns (uint128, uint32) {
+        uint256 pathLength_ = path_.length;
+
+        require(pathLength_ > 1, "UniswapV3Oracle: invalid path");
+        require(pathLength_ == fees_.length + 1, "UniswapV3Oracle: path/fee lengths do not match");
+        require(
+            block.timestamp > period_,
+            "UniswapV3Oracle: period larger than current timestamp"
+        );
+        require(period_ > 0, "UniswapV3Oracle: period can't be 0");
+
+        if (amount_ == 0) {
+            return (0, 0);
+        }
+
+        uint32 minPeriod_ = period_;
+
+        for (uint256 i = 0; i < pathLength_ - 1; i++) {
+            (uint256 price_, uint32 time_) = _getPriceOfTokenInToken(
+                path_[i],
+                path_[i + 1],
+                amount_,
+                fees_[i],
+                period_
+            );
+
+            amount_ = SafeCast.toUint128(price_);
+
+            if (minPeriod_ > time_) {
+                minPeriod_ = time_;
+            }
+        }
+
+        return (amount_, minPeriod_);
+    }
+
+    /**
+     * @notice Function to get timestamp from the oldest available observation
+     */
+    function _findOldestObservation(address pool_) internal view returns (uint32) {
+        (, , uint16 observationIndex_, uint16 observationCardinality_, , , ) = IUniswapV3Pool(
+            pool_
+        ).slot0();
+
+        require(observationCardinality_ > 0, "UniswapV3Oracle: pool is not initialized");
+
+        (uint32 blockTimestamp_, , , bool initialized_) = IUniswapV3Pool(pool_).observations(
+            (observationIndex_ + 1) % observationCardinality_
+        );
+
+        if (!initialized_) {
+            (blockTimestamp_, , , ) = IUniswapV3Pool(pool_).observations(0);
+        }
+
+        return blockTimestamp_;
+    }
+
+    /**
+     * @notice The private function to get the price of a token inside a pool
+     * @dev Returns price multplied by 10 in power of decimals of the quoteToken_
+     */
+    function _getPriceOfTokenInToken(
+        address baseToken_,
+        address quoteToken_,
+        uint128 amount_,
+        uint24 fee_,
+        uint32 period_
+    ) private view returns (uint256, uint32) {
+        if (baseToken_ == quoteToken_) {
+            return (amount_, period_);
+        }
+
+        address pool_ = uniswapV3Factory.getPool(baseToken_, quoteToken_, fee_);
+
+        require(pool_ != address(0), "UniswapV3Oracle: such pool doesn't exist");
+
+        uint32 oldest_ = _findOldestObservation(pool_); //oldest available timestamp
```

Comment:
Let's add this function to TickHelper, since it turned out uni v3 implements it. 
Name this var `longestPeriod_` (in seconds)

Reply:
But this var is not really period, it's the oldest available timestamp (that's the little difference between uni v3 function and this)

Reply:
Maybe, oldestTimestamp_ then?

Reply:
Agree, I mean let's just take the implementation from uni which returns seconds

Reply:
Okay, resolved